saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs

```diff
@@ -111,13 +113,30 @@ groundToFOV BaseRealRepr    _         = Left "Real is not FOV"
 groundToFOV BaseComplexRepr         _ = Left "Complex is not FOV"
 groundToFOV (BaseStringRepr _)      _ = Left "String is not FOV"
 groundToFOV (BaseFloatRepr _)       _ = Left "Floating point is not FOV"
-groundToFOV (BaseArrayRepr (Empty :> ty) b) _
-  | Right fot1 <- typeReprToFOT ty
-  , Right fot2 <- typeReprToFOT b
-  = pure $ FOVArray fot1 fot2
-groundToFOV (BaseArrayRepr _idx _b) _ = Left "Unsupported FOV Array"
+groundToFOV (BaseArrayRepr _ _) (ArrayMapping _) =
+    Left "Unsupported FOV Array (values only available by lookup)"
+groundToFOV (BaseArrayRepr (Empty :> ty_idx) ty_val) (ArrayConcrete dfl values) = do
+    ty_idx' <- typeReprToFOT ty_idx
+    dfl' <- groundToFOV ty_val dfl
+    let convert mvalues' idx v = do
+          values' <- mvalues'
+          idx' <- indexToFOV ty_idx idx
+          v' <- groundToFOV ty_val v
+          return $ Map.insert idx' v' values'
+    values' <- Map.foldlWithKey convert (pure Map.empty) values
+    pure $ FOVArray ty_idx' dfl' values'
+groundToFOV (BaseArrayRepr _ _) _ =
+    Left "Unsupported FOV array (unexpected key type)"
 groundToFOV (BaseStructRepr ctx) tup  = FOVTuple <$> tupleToList ctx tup
 
+indexToFOV :: BaseTypeRepr ty -> Assignment IndexLit (EmptyCtx ::> ty) -> Either String FirstOrderValue
+indexToFOV BaseIntegerRepr (Empty :> (IntIndexLit k)) =
+   pure $ FOVInt k
+indexToFOV (BaseBVRepr w)  (Empty :> (BVIndexLit _w k)) =
+   pure $ FOVWord (natValue w) (BV.asUnsigned k)
```

Comment:
Another assumption: we are assuming that the index type will either be an integer or a bitvector. Again, I think that's fine, given that this is a requirement imposed by the `what4` API.

Reply:
> I tried adjusting the patterns so that it would fail to compile if What4 grows more cases for index types, but it doesn't work; it demands a case for e.g. BaseBoolRepr even though there are no constructors of IndexLit that can produce that. Seems like a GHC bug.

That part is expected behavior. It's not complaining about the `IndexLit`, but rather the possibility of matching on other constructors of `BaseTypeRepr`. For example, you could conceivably reach the error case of this function if you passed something besides `BaseIntegerRepr` or `BaseBVRepr`, e.g.,

```
> indexToFOV BaseRealRepr (Ctx.Empty Ctx.:> undefined)
*** Exception: indexToFOV: unexpected array index type
CallStack (from HasCallStack):
  error, called at src/Verifier/SAW/Simulator/What4/FirstOrder.hs:142:5 in saw-core-what4-0.2-inplace:Verifier.SAW.Simulator.What4.FirstOrder
```

Reply:
Having typed all that up, it occurs to me that you don't really need the `BaseTypeRepr` at all. Currently, the only reason to match against a `BaseTypeRepr` is to extract the underlying bitvector width `w` from a `BaseBVRepr`, but that same width value is also stored in a `BVIndexLit`. Therefore, I propose that you get rid of the `BaseTypeRepr` argument entirely:

```hs
indexToFOV :: Assignment IndexLit (EmptyCtx ::> ty) -> Either String FirstOrderValue
indexToFOV (Empty :> idxLit) =
  case idxLit of
    IntIndexLit k -> pure $ FOVInt k
    BVIndexLit w k -> pure $ FOVWord (natValue w) (BV.asUnsigned k)
```

Reply:
Re pattern matching, I see, it isn't actually matching pairs of arguments at once.

But yeah, may as well just drop the type. And in that case it doesn't need to return an Either, either :-)

Reply:
Oops: fixed in 74475a2206612df1ba9b384611aa25cde3f347e0